File: src/leetcode/python/medium/PathSumII_113.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class PathSumII_113(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root:
            return []
        if not root.left and not root.right:
            if root.val == sum:
                return [[root.val]]
            else:
                return []

        tmp = self.pathSum(root.left, sum-root.val) + \
            self.pathSum(root.right, sum-root.val)
        return [[root.val] + i for i in tmp]

File: src/leetcode/java/PathSumII_113.java
package leetcode.java;

import java.util.ArrayList;
import java.util.List;

public class PathSumII_113 {
  private List<List<Integer>> lists = new ArrayList<>();
  private List<Integer> list = new ArrayList<>();
  private int required;

  public List<List<Integer>> pathSum(TreeNode root, int sum) {
    required = sum;
    if (root == null) return lists;
    if (root.left == null && root.right == null) {
      if (root.val == required) {
        list.add(root.val);
        lists.add(list);
      }
      return lists;
    }
    list.add(root.val);
    dfs(root.left);
    dfs(root.right);
    return lists;
  }

  private void dfs(TreeNode node) {
    if (node == null) return;
    if (node.left == null && node.right == null) {
      if (caculateLeafSum(node.val) == required) {
        list.add(node.val);
        lists.add(new ArrayList<>(list));
        list.remove(list.size() - 1);
      }
      return;
    }
    list.add(node.val);
    dfs(node.left);
    dfs(node.right);
    list.remove(list.size() - 1);
  }

  private int caculateLeafSum(int val) {
    int sum = val;
    for (int ele : list) sum += ele;
    return sum;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
